hack/generate/changelog/gen-changelog.go

```diff
@@ -0,0 +1,331 @@
+package main
+
+import (
+	"bytes"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"gopkg.in/yaml.v2"
+
+	log "github.com/sirupsen/logrus"
+)
+
+func main() {
+	var (
+		tag           string
+		fragmentsDir  string
+		changelogFile string
+		migrationDir  string
+		validateOnly  bool
+	)
+
+	flag.StringVar(&tag, "tag", "",
+		"Title for generated CHANGELOG and migration guide sections")
+	flag.StringVar(&fragmentsDir, "fragments-dir", filepath.Join("changelog", "fragments"),
+		"Path to changelog fragments directory")
+	flag.StringVar(&changelogFile, "changelog", "CHANGELOG.md",
+		"Path to CHANGELOG")
+	flag.StringVar(&migrationDir, "migration-guide-dir",
+		filepath.Join("website", "content", "en", "docs", "migration"),
+		"Path to migration guide directory")
+	flag.BoolVar(&validateOnly, "validate-only", false,
+		"Only validate fragments")
+	flag.Parse()
+
+	if tag == "" && !validateOnly {
+		log.Fatalf("flag '-tag' is required without '-validate-only'")
+	}
+
+	entries, err := loadEntries(fragmentsDir)
+	if err != nil {
+		log.Fatalf("failed to load fragments: %v", err)
+	}
+	if len(entries) == 0 {
+		log.Fatalf("no Entries found")
+	}
+
+	if validateOnly {
+		return
+	}
+
+	if err := updateChangelog(config{
+		File:    changelogFile,
+		Title:   tag,
+		Entries: entries,
+	}); err != nil {
+		log.Fatalf("failed to update CHANGELOG: %v", err)
+	}
+
+	if err := createMigrationGuide(config{
+		File:    filepath.Join(migrationDir, fmt.Sprintf("%s.md", tag)),
+		Title:   tag,
+		Entries: entries,
+	}); err != nil {
+		log.Fatalf("failed to update migration guide: %v", err)
+	}
+}
+
+func loadEntries(fragmentsDir string) ([]entry, error) {
+	files, err := ioutil.ReadDir(fragmentsDir)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read fragments directory: %w", err)
+	}
+
+	var entries []entry
+	for _, fragFile := range files {
+		if fragFile.Name() == "00-template.yaml" {
+			continue
+		}
+		if fragFile.IsDir() {
+			log.Warnf("Skipping directory %q", fragFile.Name())
+			continue
+		}
+		if filepath.Ext(fragFile.Name()) != ".yaml" || fragFile.IsDir() {
```

Comment:
Already being checked above.
```suggestion
		if filepath.Ext(fragFile.Name()) != ".yaml" {
```